### ISSUE TITLE
"Unable to seek to sector" regression

### ISSUE THREAD

**Issue by jennybc** [evanmiller/libxls#17]
The release of readxl v1.1.0 has brought a regression to light. Related issue in readxl https://github.com/tidyverse/readxl/issues/463.

I can reproduce the OP's report that this xls file could be read with readxl v1.0.0 but not with readxl v1.1.0.

I can also reproduce the failure with xls2csv built from current HEAD here (79d31e0e7c7dfcc93aec6bb0accb2128856c2c79), i.e. without readxl in the equation. I *think* it's a coincidence that I managed to break things for the same file on the xlsx side as well 🤔. Rather, I imagine this tool (Economist Intelligent Unit) writes valid-but-nonstandard files for both xls and xlsx.

```
jenny@2015-mbp libxls-evanmiller-github $ git log --pretty=oneline -1
79d31e0e7c7dfcc93aec6bb0accb2128856c2c79 (HEAD -> master, upstream/master) Merge pull request #16 from jennybc/simplify-uint64
jenny@2015-mbp libxls-evanmiller-github $ ./xls2csv ~/Downloads/EIU/EIU_Data.xls
FILE: /Users/jenny/Downloads/EIU/EIU_Data.xls
Error reading XLS file: Unable to open file
```

I know you are mostly focused on other things here, at the moment. But libxls used to read this file OK. I'm attaching the OP's xls in a `.zip` file.
[EIU.zip](https://github.com/evanmiller/libxls/files/1940127/EIU.zip)

**Comment by jimhester** [comment]
Here is the git bisect log, 81d0eaee00903713909dec895625a16bc0b7686e seems to be the first commit that fails to read `EIU_Data.xls` linked above.

```
git bisect start
# bad: [79d31e0e7c7dfcc93aec6bb0accb2128856c2c79] Merge pull request #16 from jennybc/simplify-uint64
git bisect bad 79d31e0e7c7dfcc93aec6bb0accb2128856c2c79
# good: [087219602c17dfedaa0568916a0cd96d60015a0f] Redo numerics
git bisect good 087219602c17dfedaa0568916a0cd96d60015a0f
# good: [58a33b9bf49954de25758282cd21b2ac707b086f] Fix out-of-bounds memory error
git bisect good 58a33b9bf49954de25758282cd21b2ac707b086f
# bad: [9cd2db7ef3d5a189f65f009501d9daf440352001] Fix more memory errors thanks to fuzzing
git bisect bad 9cd2db7ef3d5a189f65f009501d9daf440352001
# good: [ebeb54290da17f7c0abf2175d5bdeddcdd73fd2d] Appveyor fixes
git bisect good ebeb54290da17f7c0abf2175d5bdeddcdd73fd2d
# good: [ce8e8b159a0df29db177865a00f3a3e9a3ef34d7] Basic fuzzing facilities
git bisect good ce8e8b159a0df29db177865a00f3a3e9a3ef34d7
# bad: [a43f88c5f6c62c67abc93ef74d14e4379f9dc844] Fix memory errors found with fuzzing
git bisect bad a43f88c5f6c62c67abc93ef74d14e4379f9dc844
# bad: [81d0eaee00903713909dec895625a16bc0b7686e] Fix memory out-of-bounds errors
git bisect bad 81d0eaee00903713909dec895625a16bc0b7686e
# good: [878bad6196966171c81272b244cc2671ee9ab2b5] Conditionally print errors
git bisect good 878bad6196966171c81272b244cc2671ee9ab2b5
# first bad commit: [81d0eaee00903713909dec895625a16bc0b7686e] Fix memory out-of-bounds errors
```

**Comment by evanmiller** [comment]
Based on the bisect information, and the nature of the offending patch, it appears that this file caused a buffer overrun in previous versions of libxls. I've turned on debugging messages in `xls2csv` to assist with the debugging process.

**Comment by jennybc** [comment]
Here's another regression (readable with readxl v1.0.0 but not with readxl v1.10 and not with xls2csv build from current libxls).

https://www2.census.gov/programs-surveys/susb/tables/1999/county_naicssector_emplchange_1998-1999.xls

Related readxl issue https://github.com/tidyverse/readxl/issues/466

**Comment by evanmiller** [comment]
Thanks for the additional file. It appears to be the same bug affecting both files:

```
$ ./xls2csv ~/Downloads/EIU/EIU_Data.xls
FILE: /Users/emiller/Downloads/EIU/EIU_Data.xls
ole2_open: /Users/emiller/Downloads/EIU/EIU_Data.xls
Error: Unable to seek to sector #35905
File "/Users/emiller/Downloads/EIU/EIU_Data.xls" not found
Error reading XLS file: Unable to open file
```

```
$ ./xls2csv ~/Downloads/county_naicssector_emplchange_1998-1999.xls
FILE: /Users/emiller/Downloads/county_naicssector_emplchange_1998-1999.xls
ole2_open: /Users/emiller/Downloads/county_naicssector_emplchange_1998-1999.xls
Error: Unable to seek to sector #13934
File "/Users/emiller/Downloads/county_naicssector_emplchange_1998-1999.xls" not found
Error reading XLS file: Unable to open file
```

The error occurs when libxls attempts to write memory beyond the amount allocated. My attempts to provide naive fixes result in other errors.

**Comment by evanmiller** [comment]
I was able to track down the issue; it appears that the out-of-bounds error was introduced in https://github.com/evanmiller/libxls/commit/61a22747ac3e5328d54a0287274bd8956d9e74f5. The limit of 109 sectors should only be applied to the entries appearing in the file header, but it was also applied to the total allocations. When bounds checks were added for `SecID`, attempts to write past the 109th sector resulted in an error. In any event, I have fixed the logic.

Please re-open if you continue to see the same "Unable to seek to sector" error message. For other regressions, please open a new issue.

**Comment by jennybc** [comment]
Nice work! I just patched readxl to c847d632956bde62e51bf2d75f397d1535c6aac6 (or actually one commit later, dd658814c97d453f592ea5ec0d6971344d1cc10c) and everything looks good. Specifically, I can once again import the two files above. I'll make a formal PR to see travis/appveyor results, then merge. But unless I speak up, assume this is 👍 on the readxl front.
